In [1]:
import pandas as pd
import json
from datetime import datetime
from zoneinfo import ZoneInfo

Check duplicates for User and Course Info

In [ ]:
with open('../data_extraction/user_course_info.json', 'r') as file:
    user_course_info_check = json.load(file)

df_user_course_info_check = pd.DataFrame(user_course_info_check)

In [3]:
df_user_course_info_check

,user_id,nickname,is_tutor,is_pro,origin_country_id,living_country_id,origin_city_id,origin_city_name,living_city_id,living_city_name,timezone,trial_length,has_trial,trial_price,min_price,trial_session_count,has_beginner_course
0,5467830,⭐Learn with Kim/Kin,1,0,CN,KR,CN00001,Shanghai,KR00001,Seoul,Asia/Shanghai,2,0,1988,500,562,1
1,11401921,Teacher Emee,1,0,PH,PH,PH00000,Other,PH00000,Other,Asia/Manila,2,0,500,500,186,1
2,9627036,Shyam Syangtan,1,0,IN,IN,IN00771,Sonipat,IN00231,Delhi,Asia/Kolkata,2,0,700,500,1366,1
3,5787377,Kimi(y)a 👩🏻🎓❤️,1,0,IR,IT,IR00016,Tabriz,IT00016,Turin,Asia/Tehran,2,0,799,500,384,1
4,9934037,Anne Rola,1,0,PH,PH,PH00000,Other,PH00236,Tacloban,Asia/Shanghai,2,0,500,500,115,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11164,8392625,Yunhee Jung,1,0,US,US,US00346,East Los Angeles,US00346,East Los Angeles,America/Los_Angeles,2,0,1500,5000,0,1
11165,9170372,Hera 최해라,1,1,KR,GB,KR00166,Gunpo,GB00001,London,Europe/London,2,0,1500,5000,291,1
11166,23372570,MyKoreanTeacherDan,0,1,NZ,NZ,NZ00000,Other,NZ00001,Auckland,Pacific/Auckland,3,0,3500,5000,0,0
11167,8296973,Leeji 리지(Ji Eun Lee),0,1,KR,DE,KR00001,Seoul,DE00101,Karlsruhe,Europe/Berlin,2,0,2000,5000,0,1


In [4]:
list(df_user_course_info_check.columns)

['user_id',
 'nickname',
 'is_tutor',
 'is_pro',
 'origin_country_id',
 'living_country_id',
 'origin_city_id',
 'origin_city_name',
 'living_city_id',
 'living_city_name',
 'timezone',
 'trial_length',
 'has_trial',
 'trial_price',
 'min_price',
 'trial_session_count',
 'has_beginner_course']

In [5]:
columns_user = ['user_id',
 'nickname',
 'is_tutor',
 'is_pro',
 'origin_country_id',
 'living_country_id',
 'origin_city_id',
 'origin_city_name',
 'living_city_id',
 'living_city_name',
 'timezone']

Check if there are users that appear more than twice

In [6]:
df_user_dup = df_user_course_info_check[df_user_course_info_check[['user_id', 'nickname']].duplicated(keep=False)]

In [7]:
df_unique = df_user_dup.groupby(['user_id', 'nickname']).count()
# df_unique = df_unique.reset_index(names='user_id')

In [8]:
df_unique

,,is_tutor,is_pro,origin_country_id,living_country_id,origin_city_id,origin_city_name,living_city_id,living_city_name,timezone,trial_length,has_trial,trial_price,min_price,trial_session_count,has_beginner_course
user_id,nickname,,,,,,,,,,,,,,,
495195,Maria Exam Work etc,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
707465,Andrea,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
719539,Teddy,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
745006,Rose,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
755999,Tati,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
30946403,Edoardo,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
31046452,Kohl Sandra,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
31264233,Valentin Bétourné,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2


Users that have multiple entries of the same user_id and same nickname are removed.\
The latest one is kept because it means that they have updated some of their personal information

In [9]:
df_user_course_sorted = df_user_course_info_check.sort_values('user_id')
df_no_dup_id_nickname = df_user_course_sorted.drop_duplicates(subset=['user_id', 'nickname'], keep='last')
df_no_dup_id_nickname

,user_id,nickname,is_tutor,is_pro,origin_country_id,living_country_id,origin_city_id,origin_city_name,living_city_id,living_city_name,timezone,trial_length,has_trial,trial_price,min_price,trial_session_count,has_beginner_course
5833,55502,Micky Mick,1,0,CN,CN,CN00226,Fushun,CN00006,Beijing,Asia/Shanghai,2,0,1000,2000,97,1
5929,113638,Xin,1,1,CN,CA,,,CA00006,Montreal,America/Toronto,2,0,999,2500,227,0
5642,114708,Amy (Féng lǎoshī）,0,1,CN,CN,CN00201,Jilin,CN00201,Jilin,Asia/Shanghai,2,0,500,1990,6,1
10791,132815,Julie 줄리,1,0,KR,KR,KR00001,Seoul,KR00001,Seoul,Asia/Seoul,2,0,750,1500,0,1
7858,148092,Jhon Jairo,0,1,CO,CO,CO00101,Armenia,CO00006,Medellin,America/Bogota,2,0,1000,2500,182,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4655,32167939,Andrele,1,0,CA,CA,CA00000,Other,CA00000,Other,America/Toronto,2,0,1000,2000,3,1
220,32168520,Daphne👩🏫🌸💗,1,0,PH,PH,PH00141,Baguio,PH00000,Other,Asia/Manila,3,0,500,700,0,0
4152,32169200,Davide Castellani,1,0,IT,IT,IT00000,Other,IT00000,Other,Europe/Rome,4,0,1400,1400,0,1
4150,32169731,FelixRansford,1,0,GB,ES,GB00281,Exeter,ES00131,Jerez de la Frontera,Europe/Madrid,2,0,600,1400,0,0


In [10]:
# diff = df_unique.compare(df_unique_all)

In [11]:
user_course_info_cleaned = df_no_dup_id_nickname.to_dict(orient='records')

In [12]:
user_course_info_cleaned

[{'user_id': 55502,
  'nickname': 'Micky Mick',
  'is_tutor': 1,
  'is_pro': 0,
  'origin_country_id': 'CN',
  'living_country_id': 'CN',
  'origin_city_id': 'CN00226',
  'origin_city_name': 'Fushun',
  'living_city_id': 'CN00006',
  'living_city_name': 'Beijing',
  'timezone': 'Asia/Shanghai',
  'trial_length': 2,
  'has_trial': 0,
  'trial_price': 1000,
  'min_price': 2000,
  'trial_session_count': 97,
  'has_beginner_course': 1},
 {'user_id': 113638,
  'nickname': 'Xin',
  'is_tutor': 1,
  'is_pro': 1,
  'origin_country_id': 'CN',
  'living_country_id': 'CA',
  'origin_city_id': '',
  'origin_city_name': '',
  'living_city_id': 'CA00006',
  'living_city_name': 'Montreal',
  'timezone': 'America/Toronto',
  'trial_length': 2,
  'has_trial': 0,
  'trial_price': 999,
  'min_price': 2500,
  'trial_session_count': 227,
  'has_beginner_course': 0},
 {'user_id': 114708,
  'nickname': 'Amy (Féng lǎoshī）',
  'is_tutor': 0,
  'is_pro': 1,
  'origin_country_id': 'CN',
  'living_country_id': 'C

Check if 'has_trial' column in the users and trial table indicates 0 but trial count is more than 1.\
From the code block below, all users have the column 'has_trial' as 0.\
This means that if the trial price is greater than 0 or the trial session count is greater than 0, it does not matter what the 'has_trial' indicates\
Hence the has_trial column should not be used.

In [75]:
df_no_dup_id_nickname[df_no_dup_id_nickname['has_trial'] == 0].count()

user_id                10470
nickname               10470
is_tutor               10470
is_pro                 10470
origin_country_id      10470
living_country_id      10470
origin_city_id         10470
origin_city_name       10470
living_city_id         10470
living_city_name       10470
timezone               10470
trial_length           10470
has_trial              10470
trial_price            10470
min_price              10470
trial_session_count    10470
has_beginner_course    10470
dtype: int64

In [14]:
with open('./user_course_info_cleaned.json', 'w') as file:
        json.dump(user_course_info_cleaned, file, indent=4)

In [13]:
df_no_dup_id_nickname[df_no_dup_id_nickname['user_id'] == 1689306]

,user_id,nickname,is_tutor,is_pro,origin_country_id,living_country_id,origin_city_id,origin_city_name,living_city_id,living_city_name,timezone,trial_length,has_trial,trial_price,min_price,trial_session_count,has_beginner_course
5228,1689306,Cristina Domènech,0,1,ES,NL,ES00006,Barcelona,NL00111,Leiden,Europe/Amsterdam,2,0,4600,7400,1116,1


Check duplicates in teacher_stats

In [14]:
with open('../data_extraction/teacher_stats.json', 'r') as file:
    teacher_stats_check = json.load(file)

df_teacher_statscheck = pd.DataFrame(teacher_stats_check)

In [22]:
df_teacher_statscheck

,user_id,finished_session,response_rate,attendance_rate
0,5467830,4220,1.00,1.00
1,11401921,7443,1.00,1.00
2,9627036,6346,1.00,1.00
3,5787377,5007,1.00,1.00
4,9934037,2664,1.00,0.99
...,...,...,...,...
11032,8712350,1143,1.00,1.00
11033,8392625,364,0.98,0.99
11034,9170372,5033,1.00,1.00
11035,23372570,12,1.00,1.00


In [15]:
df_teacher_stats_dup = df_teacher_statscheck[df_teacher_statscheck[['user_id']].duplicated(keep=False)]

In [16]:
df_teacher_stats_dup

,user_id,finished_session,response_rate,attendance_rate
0,5467830,4220,1.00,1.00
5,7023372,4547,1.00,1.00
76,28084220,269,1.00,1.00
93,10599672,1068,1.00,1.00
98,7603855,404,0.99,0.98
...,...,...,...,...
10945,3524894,7374,1.00,1.00
10951,11663378,1443,1.00,1.00
10952,5728578,4621,1.00,1.00
10956,815733,3168,0.99,1.00


In [17]:
df_teacher_stats_count = df_teacher_stats_dup.groupby('user_id').count()
df_teacher_stats_count

,finished_session,response_rate,attendance_rate
user_id,,,
222305,3,3,3
495195,2,2,2
502096,2,2,2
706352,2,2,2
707465,2,2,2
...,...,...,...
30739843,2,2,2
30934061,2,2,2
31068665,2,2,2


In [18]:
df_teacher_stats_count[df_teacher_stats_count['finished_session'] >= 4].count()

finished_session    0
response_rate       0
attendance_rate     0
dtype: int64

In [19]:
df_teacher_stats_sorted = df_teacher_statscheck.sort_values('user_id')
df_teacher_stats_no_dup = df_teacher_stats_sorted.drop_duplicates(subset=['user_id'], keep='last')
df_teacher_stats_no_dup

,user_id,finished_session,response_rate,attendance_rate
5838,55502,2689,1.00,1.00
5933,113638,8410,1.00,1.00
5648,114708,50,0.94,0.98
10675,132815,2046,1.00,1.00
7838,148092,4109,1.00,1.00
...,...,...,...,...
4659,32167939,3,1.00,1.00
220,32168520,1,1.00,1.00
4152,32169200,0,0.00,0.00
4150,32169731,0,0.00,0.00


In [20]:
df_teacher_statscheck
df_teacher_statscheck[df_teacher_statscheck['user_id'] == 1689306]

,user_id,finished_session,response_rate,attendance_rate
4077,1689306,15479,1.0,1.0
5247,1689306,15488,1.0,1.0


In [21]:
df_teacher_stats_no_dup[df_teacher_stats_no_dup['user_id'] == 1689306]

,user_id,finished_session,response_rate,attendance_rate
5247,1689306,15488,1.0,1.0


In [22]:
teacher_stats_cleaned = df_teacher_stats_no_dup.to_dict(orient='records')

In [23]:
teacher_stats_cleaned

[{'user_id': 55502,
  'finished_session': 2689,
  'response_rate': 1.0,
  'attendance_rate': 1.0},
 {'user_id': 113638,
  'finished_session': 8410,
  'response_rate': 1.0,
  'attendance_rate': 1.0},
 {'user_id': 114708,
  'finished_session': 50,
  'response_rate': 0.94,
  'attendance_rate': 0.98},
 {'user_id': 132815,
  'finished_session': 2046,
  'response_rate': 1.0,
  'attendance_rate': 1.0},
 {'user_id': 148092,
  'finished_session': 4109,
  'response_rate': 1.0,
  'attendance_rate': 1.0},
 {'user_id': 222305,
  'finished_session': 14061,
  'response_rate': 1.0,
  'attendance_rate': 1.0},
 {'user_id': 249152,
  'finished_session': 1249,
  'response_rate': 0.97,
  'attendance_rate': 0.99},
 {'user_id': 259649,
  'finished_session': 8798,
  'response_rate': 1.0,
  'attendance_rate': 0.99},
 {'user_id': 291538,
  'finished_session': 2503,
  'response_rate': 0.98,
  'attendance_rate': 0.98},
 {'user_id': 316582,
  'finished_session': 5608,
  'response_rate': 1.0,
  'attendance_rate': 0

In [24]:
with open('./teacher_stats_cleaned.json', 'w') as file:
        json.dump(teacher_stats_cleaned, file, indent=4)

Check if tutors have no pro course

In [25]:
with open('../data_extraction/pro_course.json', 'r') as file:
    pro_course_check = json.load(file)

df_pro_course_check = pd.DataFrame(pro_course_check)

In [26]:
df_pro_course_count = df_pro_course_check.groupby('teacher_id').count()
df_pro_course_count

,id,language,title,session_price,student_count,session_count,has_package
teacher_id,,,,,,,
55502,3,3,3,3,3,3,3
113638,4,4,4,4,4,4,4
114708,4,4,4,4,4,4,4
132815,5,5,5,5,5,5,5
148092,4,4,4,4,4,4,4
...,...,...,...,...,...,...,...
32167939,2,2,2,2,2,2,2
32168520,3,3,3,3,3,3,3
32169200,1,1,1,1,1,1,1


In [27]:
df_pro_course_count[df_pro_course_count['id'] > 0]

,id,language,title,session_price,student_count,session_count,has_package
teacher_id,,,,,,,
55502,3,3,3,3,3,3,3
113638,4,4,4,4,4,4,4
114708,4,4,4,4,4,4,4
132815,5,5,5,5,5,5,5
148092,4,4,4,4,4,4,4
...,...,...,...,...,...,...,...
32167939,2,2,2,2,2,2,2
32168520,3,3,3,3,3,3,3
32169200,1,1,1,1,1,1,1


Pro Course Cleaned JSON generated below when checking if there are duplicated pro courses

Check for any duplicates in the also speaks file\
Check if there are teachers in the 'also_speaks' file but missing in the 'user_course_info' file.

In [28]:
with open('../data_extraction/also_speaks.json', 'r') as file:
    also_speaks_check = json.load(file)

df_also_speaks = pd.DataFrame(also_speaks_check)

No exact duplicates found in the also speaks dataframe.\
Exact duplicates need to be checked because with the user_course_info table, there is a one to many relationship.\
There are teachers who speak multiple other languages so they will have multiple entries in the 'also speaks' table

In [29]:
all_duplicates = df_also_speaks[df_also_speaks.duplicated(keep=False)]
all_duplicates

,user_id,language


In [30]:
with open('./user_course_info_cleaned.json', 'r') as file:
    user_course_info = json.load(file)

df_user_course_info = pd.DataFrame(user_course_info)

In [31]:
df_num_of_languages = df_also_speaks.groupby('user_id').count().reset_index()

In [32]:
df_num_of_languages[df_num_of_languages['language'] > 0]

,user_id,language
0,55502,3
1,113638,2
2,114708,4
3,132815,2
4,148092,3
...,...,...
10289,32167939,3
10290,32168520,1
10291,32169200,1
10292,32169731,1


In [33]:
df_merged = pd.merge(df_user_course_info, df_num_of_languages, on='user_id', how='left_anti')

Doing a sample check on these teachers that are in the user_info_cleaned file but not in the also_speaks file, these 176 teachers do in fact only speak one language. Hence they do not appear in the also_speaks file.\
Using the 'right-anti' there are no teachers who appear in the also speaks file but not in the users file.

In [34]:
df_merged

,user_id,nickname,is_tutor,is_pro,origin_country_id,living_country_id,origin_city_id,origin_city_name,living_city_id,living_city_name,timezone,trial_length,has_trial,trial_price,min_price,trial_session_count,has_beginner_course,language
36,595157,Consuelo,1,1,ES,ES,ES00001,Madrid,ES00001,Madrid,Europe/Berlin,2,0,1500,2000,230,0,NaN
44,652005,Scarlet Lorraine,1,1,ZZ,US,ZZ00000,Other,US01071,Sacramento,America/Los_Angeles,2,0,700,1800,782,1,NaN
58,754979,Daniel Colman,0,1,GB,GB,ZZ00000,Other,ZZ00000,Other,Europe/Paris,2,0,500,2200,0,0,NaN
63,795706,Natalia,1,1,RU,RU,ZZ00000,Other,RU00021,Niznij Novgorod,Europe/Moscow,2,0,600,2000,128,1,NaN
89,896917,Faith Valencia,1,0,PH,PH,PH00161,San Fernando,PH00221,Olongapo,Asia/Manila,4,0,750,650,388,1,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9400,27642857,SAMANTHA KVD,1,0,KE,KE,KE00001,Nairobi,KE00001,Nairobi,Africa/Nairobi,2,0,600,800,4,1,NaN
9986,30333130,Marcus,0,1,US,MX,US01381,Winston-Salem,MX00131,Cancun,America/Cancun,2,0,500,1200,32,0,NaN
9994,30344307,Professeure Gilee,1,0,CA,CA,CA00006,Montreal,CA00006,Montreal,America/Toronto,2,0,2200,4500,1,0,NaN
10284,31409112,Mojomarty,1,0,US,TH,US00951,Philadelphia,TH00000,Other,Asia/Bangkok,2,0,600,2000,0,1,NaN


In [35]:
df_merged = pd.merge(df_user_course_info, df_num_of_languages, on='user_id', how='right_anti')

In [36]:
df_merged

,user_id,nickname,is_tutor,is_pro,origin_country_id,living_country_id,origin_city_id,origin_city_name,living_city_id,living_city_name,timezone,trial_length,has_trial,trial_price,min_price,trial_session_count,has_beginner_course,language


In [37]:
also_speaks_cleaned = df_also_speaks.to_dict(orient='records')

In [38]:
also_speaks_cleaned

[{'user_id': 5467830, 'language': 0},
 {'user_id': 5467830, 'language': 1},
 {'user_id': 5467830, 'language': 2},
 {'user_id': 11401921, 'language': 3},
 {'user_id': 11401921, 'language': 4},
 {'user_id': 11401921, 'language': 5},
 {'user_id': 9627036, 'language': 6},
 {'user_id': 9627036, 'language': 7},
 {'user_id': 5787377, 'language': 8},
 {'user_id': 5787377, 'language': 9},
 {'user_id': 5787377, 'language': 10},
 {'user_id': 5787377, 'language': 1},
 {'user_id': 5787377, 'language': 11},
 {'user_id': 5787377, 'language': 12},
 {'user_id': 5787377, 'language': 0},
 {'user_id': 9934037, 'language': 3},
 {'user_id': 9934037, 'language': 5},
 {'user_id': 7023372, 'language': 0},
 {'user_id': 7023372, 'language': 13},
 {'user_id': 7023372, 'language': 12},
 {'user_id': 7023372, 'language': 14},
 {'user_id': 7023372, 'language': 1},
 {'user_id': 7023372, 'language': 7},
 {'user_id': 7023372, 'language': 5},
 {'user_id': 8226332, 'language': 15},
 {'user_id': 8226332, 'language': 16},
 

In [112]:
with open('./also_speaks_cleaned.json', 'w') as file:
        json.dump(also_speaks_cleaned, file, indent=4)

Check that Every Pro Course is tagged to a teacher.\
There are teachers with multiple pro courses.

In [39]:
with open('./user_course_info_cleaned.json', 'r') as file:
    user_course_info = json.load(file)

df_user_course_info = pd.DataFrame(user_course_info)

In [40]:
with open('../data_extraction/pro_course.json', 'r') as file:
    pro_course_check = json.load(file)

df_pro_course_check = pd.DataFrame(pro_course_check)

In [41]:
df_pro_course_check

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
0,146280,5467830,chinese,【For new student】 Customized Lesson (A1~C2） /...,6000,260,1455,1
1,146282,5467830,chinese,【Special discount for long-term students who h...,5800,61,980,1
2,148875,5467830,chinese,(No Video & No Audio Call) Essay Correction wi...,500,14,32,0
3,186046,5467830,english,【For new student 新生课】 个人定制课程 / カスタマイズレッスン,6000,62,196,1
4,186049,5467830,english,【无需见面的作文修改】 (No Video & No Audio Call) / Ess...,500,6,8,0
...,...,...,...,...,...,...,...,...
40066,249718,23372570,korean,Korean Structure Master,5000,2,11,1
40067,249721,23372570,korean,Intensive Korean for Upper level,5000,0,0,1
40068,139499,8296973,korean,Beginner/Intermediate,5000,45,912,1
40069,139502,8296973,korean,"Academic writing (한국어 전공자, 유학생)",7000,2,17,0


The following code block shows that every course has a teacher mapped to it.

In [42]:
df_merged = pd.merge(df_user_course_info, df_pro_course_check, left_on='user_id', right_on='teacher_id', how='right_anti')
df_merged

,user_id,nickname,is_tutor,is_pro,origin_country_id,living_country_id,origin_city_id,origin_city_name,living_city_id,living_city_name,...,trial_session_count,has_beginner_course,id,teacher_id,language,title,session_price,student_count,session_count,has_package


The following code block shows that the number of teachers in the pro course table matches the user_course_info table as there are 10470 unique teachers

In [43]:
df_pro_course_check.groupby('teacher_id').count()

,id,language,title,session_price,student_count,session_count,has_package
teacher_id,,,,,,,
55502,3,3,3,3,3,3,3
113638,4,4,4,4,4,4,4
114708,4,4,4,4,4,4,4
132815,5,5,5,5,5,5,5
148092,4,4,4,4,4,4,4
...,...,...,...,...,...,...,...
32167939,2,2,2,2,2,2,2
32168520,3,3,3,3,3,3,3
32169200,1,1,1,1,1,1,1


In [44]:
df_pro_course_check['teacher_id'].nunique()

10470

There are duplicate pro course id which should not be the case.\
If there id, teacher_id and the language are the same, the later one should be taken meaning sort by id and session count.\
The higher session count will be taken

In [45]:
df_pro_course_check['id'].nunique()

39035

In [46]:
df_pro_course_check_id = df_pro_course_check.groupby('id').count().reset_index()

In [47]:
df_pro_course_check_id

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
0,375,1,1,1,1,1,1,1
1,555,1,1,1,1,1,1,1
2,946,2,2,2,2,2,2,2
3,3190,1,1,1,1,1,1,1
4,4042,1,1,1,1,1,1,1
...,...,...,...,...,...,...,...,...
39030,290784,1,1,1,1,1,1,1
39031,290785,1,1,1,1,1,1,1
39032,290788,1,1,1,1,1,1,1
39033,290792,1,1,1,1,1,1,1


In [48]:
df_pro_course_check[df_pro_course_check['id'] == 946]

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
11617,946,502096,serbian,Serbian for all levels! 100% target language l...,2300,80,1112,1
35999,946,502096,serbian,Serbian for all levels! 100% target language l...,2300,80,1113,1


There are 40071 rows in the pro course table but only 39035 unique ids.

In [49]:
all_duplicates = df_pro_course_check[df_pro_course_check.duplicated(subset=['id', 'teacher_id', 'language'], keep=False)]
all_duplicates

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
0,146280,5467830,chinese,【For new student】 Customized Lesson (A1~C2） /...,6000,260,1455,1
41,101184,7023372,russian,🇷🇺 General Course - Russian for Beginners (A0-A2),2500,96,575,1
43,101662,7023372,english,🇺🇲🇨🇦🇬🇧 General English Course (Including Engli...,3000,107,1176,1
47,258619,7023372,russian,🇷🇺 NEW! Russian for Kids and Teenagers - Русск...,2500,4,172,1
384,269540,28084220,english,Free-Form Conversation in English (For adults),690,38,203,1
...,...,...,...,...,...,...,...,...
39814,221255,5728578,korean,All Skills in Korean,4500,81,2345,1
39815,247686,5728578,korean,TOPIK II Preparation (NEW! 📚📖💯),4700,4,52,1
39847,251333,23907480,korean,Korean for Beginner to Intermediate,3000,41,848,1
39858,253792,24885652,korean,Exciting Beginner Korean Lesson! [Customized f...,4300,17,222,1


In [50]:
pro_count_dup_count = all_duplicates.groupby('id').count().reset_index()
pro_count_dup_count

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
0,946,2,2,2,2,2,2,2
1,11943,2,2,2,2,2,2,2
2,13127,2,2,2,2,2,2,2
3,13362,2,2,2,2,2,2,2
4,14210,2,2,2,2,2,2,2
...,...,...,...,...,...,...,...,...
1022,289067,2,2,2,2,2,2,2
1023,289068,2,2,2,2,2,2,2
1024,289069,2,2,2,2,2,2,2
1025,289287,2,2,2,2,2,2,2


In [51]:
pro_count_dup_count[pro_count_dup_count['teacher_id'] == 2]

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
0,946,2,2,2,2,2,2,2
1,11943,2,2,2,2,2,2,2
2,13127,2,2,2,2,2,2,2
3,13362,2,2,2,2,2,2,2
4,14210,2,2,2,2,2,2,2
...,...,...,...,...,...,...,...,...
1022,289067,2,2,2,2,2,2,2
1023,289068,2,2,2,2,2,2,2
1024,289069,2,2,2,2,2,2,2
1025,289287,2,2,2,2,2,2,2


In [52]:
df_sorted = all_duplicates.sort_values(by=['teacher_id', 'id', 'session_count'], ascending=[True, True, True])
df_sorted

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
4869,11943,222305,portuguese,Português,1499,437,2337,1
19299,11943,222305,portuguese,Português,1499,439,2341,1
4871,21847,222305,german,German,1499,582,3569,1
19300,21847,222305,german,German,1499,582,3571,1
4875,40640,222305,german,Deutsch,1499,423,1966,1
...,...,...,...,...,...,...,...,...
20448,286810,31265601,italian,(Matthew) 02-IT - Lezione personalizzata di it...,2500,10,94,1
918,288579,31823132,english,english basics: conversation practice,700,2,4,1
18422,288579,31823132,english,english basics,700,2,4,1
919,288693,31823132,french,français basique: conversation,700,11,24,1


In [53]:
df_pro_course_sorted = df_pro_course_check.sort_values(by=['id', 'teacher_id', 'session_count'], ascending=[True, True, True])
df_pro_course_no_dup = df_pro_course_sorted.drop_duplicates(subset=['id', 'teacher_id', 'language'], keep='last')
df_pro_course_no_dup

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
22570,375,436016,english,English courses新概念英语（1）,1400,15,106,0
22571,555,436016,chinese,Young Learners' Chinese,1600,5,37,1
35999,946,502096,serbian,Serbian for all levels! 100% target language l...,2300,80,1113,1
22572,3190,436016,chinese,One-on-One Customized Chinese lesson,1500,73,1391,1
14692,4042,463477,english,One-on-One tutoring: English,3200,95,1504,1
...,...,...,...,...,...,...,...,...
18747,290784,31481128,french,🗣️ Natural French Conversation: Speak easily !,700,0,0,1
18455,290785,8513543,french,Leçon d'essai,800,0,0,0
38856,290788,29325526,korean,For Long-term Students,1300,0,0,1
29884,290792,6417813,spanish,DELE exam preparation / Preparación para el ex...,2700,0,0,1


In [54]:
df_pro_course_no_dup[df_pro_course_no_dup['id'] == 946]

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
35999,946,502096,serbian,Serbian for all levels! 100% target language l...,2300,80,1113,1


In [55]:
pro_course_cleaned = df_pro_course_no_dup.to_dict(orient='records')

In [56]:
pro_course_cleaned

[{'id': 375,
  'teacher_id': 436016,
  'language': 'english',
  'title': 'English courses新概念英语（1）',
  'session_price': 1400,
  'student_count': 15,
  'session_count': 106,
  'has_package': 0},
 {'id': 555,
  'teacher_id': 436016,
  'language': 'chinese',
  'title': "Young Learners' Chinese",
  'session_price': 1600,
  'student_count': 5,
  'session_count': 37,
  'has_package': 1},
 {'id': 946,
  'teacher_id': 502096,
  'language': 'serbian',
  'title': 'Serbian for all levels! 100% target language lessons!',
  'session_price': 2300,
  'student_count': 80,
  'session_count': 1113,
  'has_package': 1},
 {'id': 3190,
  'teacher_id': 436016,
  'language': 'chinese',
  'title': 'One-on-One Customized Chinese lesson',
  'session_price': 1500,
  'student_count': 73,
  'session_count': 1391,
  'has_package': 1},
 {'id': 4042,
  'teacher_id': 463477,
  'language': 'english',
  'title': 'One-on-One tutoring: English',
  'session_price': 3200,
  'student_count': 95,
  'session_count': 1504,
  'ha

In [174]:
with open('./pro_course_cleaned.json', 'w') as file:
        json.dump(pro_course_cleaned, file, indent=4)

Check if there are duplicated price list items\
Check if every price list item is linked to a course

In [57]:
with open('../data_extraction/price_list.json', 'r') as file:
    price_list = json.load(file)

df_price_list = pd.DataFrame(price_list)

In [58]:
with open('./pro_course_cleaned.json', 'r') as file:
    pro_course_cleaned = json.load(file)

df_pro_course_cleaned = pd.DataFrame(pro_course_cleaned)

In [59]:
df_price_list

,package_price,session_price,course_id,package_length,session_length,course_price_id
0,15200,3200,146280,5,2,353199
1,22000,4600,146280,5,3,353200
2,30000,6000,146280,5,4,353198
3,59000,6000,146280,10,4,1017005
4,14300,3000,146282,5,2,353214
...,...,...,...,...,...,...
216257,25000,5000,139499,5,4,333866
216258,67500,5000,139499,15,4,790345
216259,7000,7000,139502,1,4,333870
216260,25000,5000,139510,5,4,745677


There are duplicated price list items

In [60]:
df_price_list['course_price_id'].nunique()

215827

There are 435 duplicated course_price_ids and there are no course_price_ids that occur 3 times or more.\
From the sample check, the duplicates arise because of the teacher updating the prices or package/session length

In [61]:
df_price_list_dup = df_price_list.groupby('course_price_id').count()
df_price_list_dup.reset_index()
df_price_list_dup[df_price_list_dup['package_price'] > 1]

,package_price,session_price,course_id,package_length,session_length
course_price_id,,,,,
137868,2,2,2,2,2
137869,2,2,2,2,2
145354,2,2,2,2,2
151723,2,2,2,2,2
151727,2,2,2,2,2
...,...,...,...,...,...
1082724,2,2,2,2,2
1082725,2,2,2,2,2
1082726,2,2,2,2,2


In [62]:
all_duplicates = df_price_list[df_price_list.duplicated(subset=['course_price_id'], keep=False)]
all_duplicates

,package_price,session_price,course_id,package_length,session_length,course_price_id
5951,2600,600,288693,5,3,1065695
5952,5500,600,288693,10,3,1072763
5954,3000,700,288693,5,4,1065693
5956,10000,700,288693,15,4,1072764
5958,4000,900,290681,5,4,1083636
...,...,...,...,...,...,...
201593,7000,1400,274689,5,4,1075021
201594,14000,1400,274689,10,4,1075022
206621,23999,4999,107142,5,4,237738
206622,47999,4999,107142,10,4,793758


In [63]:
all_duplicates[all_duplicates['course_price_id'] == 1083636]

,package_price,session_price,course_id,package_length,session_length,course_price_id
5958,4000,900,290681,5,4,1083636
104896,4300,900,290681,5,4,1083636


In [64]:
df_price_list_sorted = df_price_list.sort_values(by=['course_price_id', 'package_price', 'session_price'], ascending=[True, True, True])
df_price_list_no_dup = df_price_list_sorted.drop_duplicates(subset=['course_price_id', 'course_id'], keep='last')
df_price_list_no_dup

,package_price,session_price,course_id,package_length,session_length,course_price_id
127459,1100,1100,375,1,3,336
127461,10000,1000,555,10,2,494
67410,20900,2300,946,10,4,823
127467,12000,1300,3190,10,3,3075
127465,10000,1000,3190,10,2,3076
...,...,...,...,...,...,...
140883,10200,1100,287665,10,3,1084998
140884,15000,1100,287665,15,3,1084999
140885,20000,1100,287665,20,3,1085000
204373,59999,7498,87801,10,4,1085029


In [65]:
df_pro_course_cleaned

,id,teacher_id,language,title,session_price,student_count,session_count,has_package
0,375,436016,english,English courses新概念英语（1）,1400,15,106,0
1,555,436016,chinese,Young Learners' Chinese,1600,5,37,1
2,946,502096,serbian,Serbian for all levels! 100% target language l...,2300,80,1113,1
3,3190,436016,chinese,One-on-One Customized Chinese lesson,1500,73,1391,1
4,4042,463477,english,One-on-One tutoring: English,3200,95,1504,1
...,...,...,...,...,...,...,...,...
39030,290784,31481128,french,🗣️ Natural French Conversation: Speak easily !,700,0,0,1
39031,290785,8513543,french,Leçon d'essai,800,0,0,0
39032,290788,29325526,korean,For Long-term Students,1300,0,0,1
39033,290792,6417813,spanish,DELE exam preparation / Preparación para el ex...,2700,0,0,1


When the pro course dataframe is joined with the cleaned price list data frame, both left-anti and right-anti join do not return any rows.\
This means that there are no ids that are not mapped between the two tables.\
All the prices are linked to a course.

In [66]:
df_merged = pd.merge(df_pro_course_cleaned, df_price_list_no_dup, left_on='id', right_on='course_id', how='right_anti')
# df_merged = pd.merge(df_pro_course_cleaned, df_price_list_no_dup, left_on='id', right_on='course_id', how='left_anti')
df_merged

,id,teacher_id,language,title,session_price_x,student_count,session_count,has_package,package_price,session_price_y,course_id,package_length,session_length,course_price_id


In [68]:
price_list_cleaned = df_price_list_no_dup.to_dict(orient='records')

In [69]:
price_list_cleaned

[{'package_price': 1100,
  'session_price': 1100,
  'course_id': 375,
  'package_length': 1,
  'session_length': 3,
  'course_price_id': 336},
 {'package_price': 10000,
  'session_price': 1000,
  'course_id': 555,
  'package_length': 10,
  'session_length': 2,
  'course_price_id': 494},
 {'package_price': 20900,
  'session_price': 2300,
  'course_id': 946,
  'package_length': 10,
  'session_length': 4,
  'course_price_id': 823},
 {'package_price': 12000,
  'session_price': 1300,
  'course_id': 3190,
  'package_length': 10,
  'session_length': 3,
  'course_price_id': 3075},
 {'package_price': 10000,
  'session_price': 1000,
  'course_id': 3190,
  'package_length': 10,
  'session_length': 2,
  'course_price_id': 3076},
 {'package_price': 30500,
  'session_price': 3200,
  'course_id': 4042,
  'package_length': 10,
  'session_length': 4,
  'course_price_id': 3934},
 {'package_price': 24300,
  'session_price': 2700,
  'course_id': 5620,
  'package_length': 10,
  'session_length': 3,
  'cour

In [70]:
with open('./price_list_cleaned.json', 'w') as file:
        json.dump(price_list_cleaned, file, indent=4)